java/ql/src/experimental/Security/CWE/CWE-348/UseOfLessTrustedSourceLib.qll

```diff
@@ -0,0 +1,88 @@
+import java
+import DataFlow
+import semmle.code.java.dataflow.DataFlow
+import experimental.semmle.code.java.Logging
+
+/** A data flow source of the value of the `x-forwarded-for` field in the `header`. */
+class UseOfLessTrustedSource extends DataFlow::Node {
+  UseOfLessTrustedSource() {
+    exists(MethodAccess ma |
+      ma.getMethod().hasName("getHeader") and
+      ma.getArgument(0).toString().toLowerCase() = "\"x-forwarded-for\"" and
+      ma = this.asExpr()
+    )
+  }
+}
+
+/** A data flow sink of method return or log output or local print. */
+class UseOfLessTrustedSink extends DataFlow::Node {
+  UseOfLessTrustedSink() {
+    exists(ReturnStmt rs | rs.getResult() = this.asExpr())
+    or
+    exists(MethodAccess ma |
+      ma.getMethod().getName() in ["print", "println"] and
+      (
+        ma.getMethod().getDeclaringType().hasQualifiedName("java.io", "PrintWriter") or
+        ma.getMethod().getDeclaringType().hasQualifiedName("java.io", "PrintStream")
+      ) and
+      ma.getAnArgument() = this.asExpr()
+    )
+    or
+    exists(LoggingCall lc | lc.getAnArgument() = this.asExpr())
+  }
+}
+
+/** A method that split string. */
+class SplitMethod extends Method {
+  SplitMethod() {
+    this.getNumberOfParameters() = 1 and
+    this.hasQualifiedName("java.lang", "String", "split")
+  }
+}
+
+/**
+ * A call to the ServletRequest.getHeader method and the argument are
+ * `wl-proxy-client-ip`/`proxy-client-ip`/`http_client_ip`/`http_x_forwarded_for`/`x-real-ip`.
+ */
+class HeaderIpCall extends MethodAccess {
+  HeaderIpCall() {
+    this.getMethod().hasName("getHeader") and
+    this.getMethod()
+        .getDeclaringType()
+        .getASupertype*()
+        .hasQualifiedName("javax.servlet", "ServletRequest") and
+    this.getArgument(0).toString().toLowerCase() in [
+        "\"wl-proxy-client-ip\"", "\"proxy-client-ip\"", "\"http_client_ip\"",
+        "\"http_x_forwarded_for\"", "\"x-real-ip\""
```

Comment:
Explain why you're using these headers for `xffIsFirstGet` but only `X-Forwarded-For` in the main query?

Reply:
This is related to the `xffIsFirstGet` predicate

Reply:
Right, but why are you checking for all these possibilities here, but only `X-Forwarded-For` in `UseOfLessTrustedSource`? What is special about that header that means you trust it less than `X-Real-IP` for instance?

Reply:
Yes, and `X-Forwarded-For` has a high usage rate.

Reply:
Why are you only considering `X-Forwarded-For` a source of error? Isn't `X-Real-IP` exactly the same? I don't understand why you consider all 5 for the purposes of finding out if this is the first check in this function, but only one for the source.

Reply:
`X-Forwarded-For` is a superposition process when used, format: `client, proxy1, proxy2`. `X-Real-IP` is a process of coverage when in use.

Reply:
Right, which is why you suppose that taking the right-most comma-separated piece of `X-Forwarded-For` might be safer than the left-most. If `X-Real-IP` doesn't get used that way then it's simply always dangerous to trust, right? But at the moment we treat it as always safe?

Reply:
Thank you for reviewing my pr.

You are right. But I still find it weird, regarding the questions here and above, I would like to hear the opinions of security-lab. Hope you can understand.